Cargo.toml

```diff
@@ -7,11 +7,13 @@ edition = "2021"
 serde = "1"
 
 # And optional ones
-indexmap = {version = "2", optional = true}
+indexmap = { version = "2", optional = true }
+hashbrown = "0.14.1"
```

Comment:
if the goal is to use ahash, we should depend on it directly rather than use hashbrown

Reply:
I'll give it a try, but when you switch Hash impls you don't get `new` and `with_capacity`.

Reply:
`hashbrown::HashMap` is strictly faster than `std::collections::HashMap`, even when switching to `AHash`.

This is mostly a 2-5% win over `AHash`, but a 17% win in Serialization speed.

We probably want to stick with `hashbrown`.

Reply:
I pushed the std + AHash branch too so you can compare them yourself.

Reply:
I thought std was using hashbrown? Is it just a matter of waiting for the rustc version to catch up to current hashbrown version?

Reply:
I thought so too, but apparently not.

Reply:
I have no differences at all between your 2 PRs with rustc==1.70

Reply:
I'll bench them again when I have time, maybe test in a different machine to validate the results. I'll remove the draft status for the other PR so you can merge it if you trust your numbers.

Reply:
I'm seeing some jitter in the perf numbers for everything but `serialize`, so `std` vs `hashbrown` might only affect serialization...

Reply:
I implemented iai benchmarks and I see `hashbrown` being positive in `serialize` (~8% less intructions), and being also improving the templates benches around 0.5-2%.

My guess is that because `hashbrown` has access to the internals of `HashMap` it can implement `Serialize` more efficiently. Also the Rust std comes precompiled(?) AFAIK, so you can't do some optimizations(?) maybe(???)

Reply:
really weird, is there still a diff with nightly?

Reply:
Didn't have nightly installed so I tried 1.70.0 instead, I'll try nightly

Reply:
# ahash vs hashbrown

Legend: `-` is a win for ahash `+` is a win for hashbrown

## Summary

- Rust 1.70: `ahash` is slightly better for big_table, otherwise `hashbrown` is equal or better.
- Rust 1.73: `hashbrown` is always better or equal.
- Rust 1.75 (nightly): `hashbrown` is always better or equal.

## Results

<details><summary>Resulst on Rust 1.70.0: (click to expand)</summary>

```log
     Running benches/iai-escape.rs (target/release/deps/iai_escape-2eaf22a40af5749d)
iai_escape::escape::bench_escape_html long_html:HTML_VERY_LONG
  Instructions:               63919 (No Change)
  L1 Hits:                    81410 (No Change)
  L2 Hits:                        3 (No Change)
  RAM Hits:                     186 (No Change)
  Total read+write:           81599 (No Change)
  Estimated Cycles:           87935 (No Change)
iai_escape::escape::bench_escape_html long_no_html:NO_HTML_VERY_LONG
  Instructions:               23598 (No Change)
  L1 Hits:                    30099 (No Change)
  L2 Hits:                        3 (No Change)
  RAM Hits:                      82 (No Change)
  Total read+write:           30184 (No Change)
  Estimated Cycles:           32984 (No Change)
iai_escape::escape::bench_escape_html short_html:HTML_SHORT
  Instructions:                2177 (No Change)
  L1 Hits:                     2837 (No Change)
  L2 Hits:                        2 (No Change)
  RAM Hits:                      42 (No Change)
  Total read+write:            2881 (No Change)
  Estimated Cycles:            4317 (No Change)
iai_escape::escape::bench_escape_html short_no_html:NO_HTML_SHORT
  Instructions:                1838 (No Change)
  L1 Hits:                     2407 (No Change)
  L2 Hits:                        2 (No Change)
  RAM Hits:                      36 (No Change)
  Total read+write:            2445 (No Change)
  Estimated Cycles:            3677 (No Change)
     Running benches/iai-templates.rs (target/release/deps/iai_templates-5012d77b3064289e)
iai_templates::templates::render teams:teams_setup()
  Instructions:               36052 (+6.210229%)
  L1 Hits:                    52159 (+5.812066%)
  L2 Hits:                       36 (+5.882353%)
  RAM Hits:                     364 (+2.247191%)
  Total read+write:           52559 (+5.786571%)
  Estimated Cycles:           65079 (+5.094955%)
iai_templates::templates::render big_table:big_table_setup()
  Instructions:            12400462 (-2.596010%)
  L1 Hits:                 18841585 (-2.236530%)
  L2 Hits:                    12260 (-0.146604%)
  RAM Hits:                    2002 (-0.298805%)
  Total read+write:        18855847 (-2.234998%)
  Estimated Cycles:        18972955 (-2.222900%)
iai_templates::templates::render realistic:realistic_setup()
  Instructions:              200772 (+4.919027%)
  L1 Hits:                   285285 (+4.825997%)
  L2 Hits:                      485 (+11.23853%)
  RAM Hits:                     729 (+0.690608%)
  Total read+write:          286499 (+4.825272%)
  Estimated Cycles:          313225 (+4.522960%)
iai_templates::templates::context big_context:big_context_setup()
  Instructions:              477963 (-0.478904%)
  L1 Hits:                   661190 (-0.754261%)
  L2 Hits:                     5709 (+18.27222%)
  RAM Hits:                     219 (+157.6471%)
  Total read+write:          667118 (-0.597353%)
  Estimated Cycles:          697400 (+0.587747%)
     Running benches/iai-value.rs (target/release/deps/iai_value-3885c5103547fdb5)
iai_value::serialize::serialize_value page:& Page :: default()
  Instructions:                6437 (+10.46851%)
  L1 Hits:                     9071 (+10.25890%)
  L2 Hits:                        6 (+50.00000%)
  RAM Hits:                     144 (+23.07692%)
  Total read+write:            9221 (+10.45759%)
  Estimated Cycles:           14141 (+14.57624%)
```

</details>

<details><summary>Results on Rust 1.73.0: (click to expand)</summary>

```log
     Running benches/iai-escape.rs (target/release/deps/iai_escape-1768f7101c765a3b)
iai_escape::escape::bench_escape_html long_html:HTML_VERY_LONG
  Instructions:               63518 (No Change)
  L1 Hits:                    81007 (No Change)
  L2 Hits:                        5 (No Change)
  RAM Hits:                     188 (No Change)
  Total read+write:           81200 (No Change)
  Estimated Cycles:           87612 (No Change)
iai_escape::escape::bench_escape_html long_no_html:NO_HTML_VERY_LONG
  Instructions:               23538 (No Change)
  L1 Hits:                    30037 (No Change)
  L2 Hits:                        4 (No Change)
  RAM Hits:                      85 (No Change)
  Total read+write:           30126 (No Change)
  Estimated Cycles:           33032 (No Change)
iai_escape::escape::bench_escape_html short_html:HTML_SHORT
  Instructions:                2149 (No Change)
  L1 Hits:                     2806 (No Change)
  L2 Hits:                        3 (No Change)
  RAM Hits:                      46 (No Change)
  Total read+write:            2855 (No Change)
  Estimated Cycles:            4431 (No Change)
iai_escape::escape::bench_escape_html short_no_html:NO_HTML_SHORT
  Instructions:                1842 (No Change)
  L1 Hits:                     2409 (No Change)
  L2 Hits:                        3 (No Change)
  RAM Hits:                      39 (No Change)
  Total read+write:            2451 (No Change)
  Estimated Cycles:            3789 (No Change)
     Running benches/iai-templates.rs (target/release/deps/iai_templates-4d7fe8e0bd8bb6d0)
iai_templates::templates::render teams:teams_setup()
  Instructions:               34103 (+2.952453%)
  L1 Hits:                    49578 (+3.382267%)
  L2 Hits:                       51 (-16.39344%)
  RAM Hits:                     368 (+1.657459%)
  Total read+write:           49997 (+3.344426%)
  Estimated Cycles:           62713 (+2.924620%)
iai_templates::templates::render big_table:big_table_setup()
  Instructions:            12267792 (+0.419135%)
  L1 Hits:                 18799746 (+1.032236%)
  L2 Hits:                    12167 (-1.065214%)
  RAM Hits:                    2000 (-0.249377%)
  Total read+write:        18813913 (+1.030712%)
  Estimated Cycles:        18930581 (+1.020554%)
iai_templates::templates::render realistic:realistic_setup()
  Instructions:              193690 (+2.888135%)
  L1 Hits:                   275484 (+3.268807%)
  L2 Hits:                      513 (-9.203540%)
  RAM Hits:                     705 (-1.260504%)
  Total read+write:          276702 (+3.230452%)
  Estimated Cycles:          302724 (+2.764963%)
iai_templates::templates::context big_context:big_context_setup()
  Instructions:              472576 (+0.081959%)
  L1 Hits:                   658757 (-0.048856%)
  L2 Hits:                     5764 (+19.56026%)
  RAM Hits:                     192 (+123.2558%)
  Total read+write:          664713 (+0.109490%)
  Estimated Cycles:          694297 (+1.180861%)
     Running benches/iai-value.rs (target/release/deps/iai_value-9ac7442b9f805117)
iai_value::serialize::serialize_value page:& Page :: default()
  Instructions:                6199 (+9.137324%)
  L1 Hits:                     8707 (+9.302034%)
  L2 Hits:                        6 (+200.0000%)
  RAM Hits:                     135 (+22.72727%)
  Total read+write:            8848 (+9.532062%)
  Estimated Cycles:           13462 (+13.83393%)
```

</details>

<details><summary>Results on Rust 1.75.0-nightly (d2f6a0485 2023-10-20): (click to expand)</summary>

```log
     Running benches/iai-escape.rs (target/release/deps/iai_escape-8ae1f368b2fb534b)
iai_escape::escape::bench_escape_html long_html:HTML_VERY_LONG
  Instructions:               59269 (No Change)
  L1 Hits:                    76386 (No Change)
  L2 Hits:                        4 (No Change)
  RAM Hits:                     186 (No Change)
  Total read+write:           76576 (No Change)
  Estimated Cycles:           82916 (No Change)
iai_escape::escape::bench_escape_html long_no_html:NO_HTML_VERY_LONG
  Instructions:               22473 (No Change)
  L1 Hits:                    28969 (No Change)
  L2 Hits:                        4 (No Change)
  RAM Hits:                      83 (No Change)
  Total read+write:           29056 (No Change)
  Estimated Cycles:           31894 (No Change)
iai_escape::escape::bench_escape_html short_html:HTML_SHORT
  Instructions:                1935 (No Change)
  L1 Hits:                     2550 (No Change)
  L2 Hits:                        3 (No Change)
  RAM Hits:                      45 (No Change)
  Total read+write:            2598 (No Change)
  Estimated Cycles:            4140 (No Change)
iai_escape::escape::bench_escape_html short_no_html:NO_HTML_SHORT
  Instructions:                1780 (No Change)
  L1 Hits:                     2344 (No Change)
  L2 Hits:                        3 (No Change)
  RAM Hits:                      37 (No Change)
  Total read+write:            2384 (No Change)
  Estimated Cycles:            3654 (No Change)
     Running benches/iai-templates.rs (target/release/deps/iai_templates-8aec8778cb7b979b)
iai_templates::templates::render teams:teams_setup()
  Instructions:               34668 (+4.048741%)
  L1 Hits:                    50198 (+4.149550%)
  L2 Hits:                       36 (+12.50000%)
  RAM Hits:                     380 (+1.876676%)
  Total read+write:           50614 (+4.137605%)
  Estimated Cycles:           63678 (+3.688144%)
iai_templates::templates::render big_table:big_table_setup()
  Instructions:            12521533 (+0.824680%)
  L1 Hits:                 19083815 (+1.291909%)
  L2 Hits:                    12545 (+2.124715%)
  RAM Hits:                    2009 (-0.149105%)
  Total read+write:        19098369 (+1.292298%)
  Estimated Cycles:        19216855 (+1.289256%)
iai_templates::templates::render realistic:realistic_setup()
  Instructions:              194983 (+3.282006%)
  L1 Hits:                   276841 (+3.507827%)
  L2 Hits:                      506 (+11.45374%)
  RAM Hits:                     731 (+1.527778%)
  Total read+write:          278078 (+3.515949%)
  Estimated Cycles:          304956 (+3.399801%)
iai_templates::templates::context big_context:big_context_setup()
  Instructions:              473095 (+0.163447%)
  L1 Hits:                   659352 (+0.020175%)
  L2 Hits:                     5728 (+18.51852%)
  RAM Hits:                     193 (+129.7619%)
  Total read+write:          665273 (+0.171200%)
  Estimated Cycles:          694747 (+1.227263%)
     Running benches/iai-value.rs (target/release/deps/iai_value-f48a236761f52427)
iai_value::serialize::serialize_value page:& Page :: default()
  Instructions:                6013 (+8.069734%)
  L1 Hits:                     8407 (+8.281813%)
  L2 Hits:                        6 (+100.0000%)
  RAM Hits:                     133 (+23.14815%)
  Total read+write:            8546 (+8.520635%)
  Estimated Cycles:           13092 (+13.26239%)
```

</details>

## Reasoning

I think having the source code of `hashbrown` is impacting performance. Usually rust ships a prebuild `std`. You can override this with `-Z build-std`.

Digging deeper by testing that assumption is inconclussive:

<details><summary>Rust nightly ahash vs ahash + build-std</summary>

```log
     Running benches/iai-escape.rs (target/x86_64-unknown-linux-gnu/release/deps/iai_escape-90a2cc40a1f09d37)
iai_escape::escape::bench_escape_html long_html:HTML_VERY_LONG
  Instructions:               59269 (No Change)
  L1 Hits:                    76384 (-0.005236%)
  L2 Hits:                        4 (+33.33333%)
  RAM Hits:                     188 (+1.621622%)
  Total read+write:           76576 (No Change)
  Estimated Cycles:           82984 (+0.127899%)
iai_escape::escape::bench_escape_html long_no_html:NO_HTML_VERY_LONG
  Instructions:               22473 (No Change)
  L1 Hits:                    28966 (-0.013807%)
  L2 Hits:                        4 (+33.33333%)
  RAM Hits:                      86 (+3.614458%)
  Total read+write:           29056 (No Change)
  Estimated Cycles:           31996 (+0.332393%)
iai_escape::escape::bench_escape_html short_html:HTML_SHORT
  Instructions:                1935 (No Change)
  L1 Hits:                     2548 (-0.156740%)
  L2 Hits:                        3 (+50.00000%)
  RAM Hits:                      47 (+6.818182%)
  Total read+write:            2598 (No Change)
  Estimated Cycles:            4208 (+2.584105%)
iai_escape::escape::bench_escape_html short_no_html:NO_HTML_SHORT
  Instructions:                1919 (No Change)
  L1 Hits:                     2515 (-0.158793%)
  L2 Hits:                        3 (+50.00000%)
  RAM Hits:                      41 (+7.894737%)
  Total read+write:            2559 (No Change)
  Estimated Cycles:            3965 (+2.746826%)
     Running benches/iai-templates.rs (target/x86_64-unknown-linux-gnu/release/deps/iai_templates-a4845cb9eafbf090)
iai_templates::templates::render teams:teams_setup()
  Instructions:               34583 (No Change)
  L1 Hits:                    50084 (-0.011978%)
  L2 Hits:                       42 (+10.52632%)
  RAM Hits:                     381 (+0.527704%)
  Total read+write:           50507 (No Change)
  Estimated Cycles:           63629 (+0.132190%)
iai_templates::templates::render big_table:big_table_setup()
  Instructions:            12521395 (No Change)
  L1 Hits:                 19083832 (-0.000686%)
  L2 Hits:                    12355 (+1.071662%)
  RAM Hits:                    2004 (No Change)
  Total read+write:        19098191 (No Change)
  Estimated Cycles:        19215747 (+0.002727%)
iai_templates::templates::render realistic:realistic_setup()
  Instructions:              194998 (+0.137626%)
  L1 Hits:                   276845 (+0.148679%)
  L2 Hits:                      495 (+1.851852%)
  RAM Hits:                     725 (-1.226158%)
  Total read+write:          278065 (+0.148026%)
  Estimated Cycles:          304695 (+0.046297%)
iai_templates::templates::context big_context:big_context_setup()
  Instructions:              473058 (-0.001057%)
  L1 Hits:                   660348 (+0.006058%)
  L2 Hits:                     4792 (-0.786749%)
  RAM Hits:                      85 (-10.52632%)
  Total read+write:          665225 (-0.001203%)
  Estimated Cycles:          687283 (-0.072697%)
     Running benches/iai-value.rs (target/x86_64-unknown-linux-gnu/release/deps/iai_value-a43dcd4ffcf6da84)
iai_value::serialize::serialize_value page:& Page :: default()
  Instructions:                6152 (+0.277099%)
  L1 Hits:                     8584 (+0.256949%)
  L2 Hits:                        5 (-28.57143%)
  RAM Hits:                     132 (No Change)
  Total read+write:            8721 (+0.229859%)
  Estimated Cycles:           13229 (+0.090792%)
```

</details>

There seems to be very minor fluctuations (probably noise dues to binary changes). `hashbrown` still shows clear wins.

Reply:
So weird. Let's go with that branch but make it optional

Reply:
> So weird. Let's go with that branch but make it optional

So a switch between hashbrown and std? 

I could make a type alias at the crate root, or have individual `cfg`s at each `use`. What do you prefer?

Reply:
probably type alias